README.md


ABOUT
===================================================================

This is a library that allows any arbitrary C++ application to connect
to a Mysql replication master and read/parse the replication binary
logs.

In effect, any application can now act like a Mysql replication slave,
without having to compile or link with any Mysql server code.

One important use-case for this library is for receiving changes in
the master database in real-time, without having the store the
master's data on the client server.

Features
-------------------------------------------------------------------
* Statistics of rps and time of every event in every table.

USAGE
===================================================================

Build requirements
-------------------------------------------------------------------

For building the library, you will need:

 * g++ with C+11 support

 * The standard mysql C client libraries (libmysqlclient).

 * The headers of the boost libraries (http://www.boost.org).
   At the minimum, you will need at least the shared_ptr.hpp, function.hpp,
   any.hpp and bind.hpp. If boost_unit_test_framework and boost_thread is
   found, tests will be built.

 * You (likely) will need to review and edit the contents of Logging.h
   and SlaveStats.h
   These headers contain the compile-time configuration of the logging
   and monitoring subsystems.
   The provided defaults are sub-standard and an example only.

Usage requirements
-------------------------------------------------------------------
 * Requires >= Mysql 5.1.23 and < Mysql 5.7.0. Tested only with some of the 5.1,
   5.5 and 5.6 versions of mysql servers. Libslave is not handling any checksum algorithms (implemented in 
   MYSQL 5.6.x), so servers must have something like "binlog_checksum=NONE"

Compiling
-------------------------------------------------------------------

Create directory "build" in source tree, step into it and run "cmake ..".
Then, if configure step is complete, run "make".

Review and edit Logging.h and SlaveStats.h to interface the library to
your app's logging and monitoring subsystems.

You can type "make test" inside of "build" directory to run tests. You
will need a working mysql server for it. Settings of mysql connection
can be adjusted in test/data/mysql.conf. Type "ctest -V" if something
went wrong and you need see test output.

Using the library
-------------------------------------------------------------------

Please see the examples in 'test/'.

You can find the programmer's API documentation on our github wiki
pages, see https://github.com/vozbu/libslave/wiki/API.

Please read this article about real project libslave usage: http://habrahabr.ru/company/mailru/blog/219015/


CREDITS
===================================================================

(c) 2011, ZAO "Begun"

https://github.com/Begun/libslave

(c) 2015, "Mail.Ru Group" LLC

https://github.com/vozbu/libslave

This library is licensed under the GNU LGPL. Please see the file LICENSE.


debian/changelog
libslave (1.2.0-1) UNRELEASED; urgency=low

  [ Egorov Anton <github.com/dohxehapo> ]
  * fixed tinyint (treated as char), short and unsigned short output
  * temporal types bugfix: TIME, TIMESTAMP and DATETIME types storage
  * temporal types from MYSQL 5.6 are now treated as temporal types from MYSQL 5.5

 -- Zakharov Sergey  <zaharov.sy@gmail.com>  Wed, 13 Apr 2016 18:45:55 +0300

libslave (1.1.0-1) UNRELEASED; urgency=low

  [ Egorov Anton <github.com/dohxehapo> ]
  * MYSQL 5.6.x support ()

 -- Zakharov Sergey  <zaharov.sy@gmail.com>  Wed, 26 Aug 2015 22:00:00 +0300

libslave (1.0.8-1) UNRELEASED; urgency=low

  * fixed gcc 4.7.2 compilation errors
  * Slave.h can be included into projects without c++11 support

 -- Zakharov Sergey <zaharov.sy@gmail.com>  Tue, 25 Aug 2015 13:06:07 +0300

libslave (1.0.7-1) UNRELEASED; urgency=low

  [ Valeriy Bykov ]
  * [*] consistent DefaultExtState
  * [*] fixed setMasterInfo and EmptyExtState.
  * [+] tests

  [ alheio ]
  * cmake + mysql
  * report host-user-pwd settings
  * report user-host-pwd

  [ tkatchev ]
  * Fix for SET fields with more than 32 elements.

  [ Valeriy Bykov ]
  * [*] using tyoes defined in slave/types.h in tests
  * [+] tests for SET field
  * [*] fixed slave_test arguments parsing
  * [*] fixed unit_test linking
  * [*] don't build dynamic library if libmysqlclient is static
  * [*] fixed error message
  * [+] db_filler
  * [+] test_client can walk from one binlog position to another and do benchmarks

  [ Victor Luchits ]
  * Pass CLIENT_REMEMBER_OPTIONS to mysql_real_connect

  [ Valeriy Bykov ]
  * [*] fixed BIT type value for gcc-4.9.2

  [ Artem Shmelev ]
  * [+] unit_test: added INSERT/UPDATE/DELETE tests
  * [+] added event_filters and massive unit_tests improvements
  * [+] added libslave stats and tests on it

  [ artshmelev ]
  * [*] translate rest of comments into English

  [ Valeriy Bykov ]
  * [!] documentation is updated

  [ Zakharov Sergey ]
  * debian

  [ Artem Shmelev ]
  * [*] fixed libslave stats

  [ sz ]

 -- alheio <zaharov.sy@gmail.com>  Fri, 21 Aug 2015 18:48:51 +0300

libslave (1.0.6-1) UNRELEASED; urgency=low

  * warnings removed

 -- alheio <zaharov.sy@gmail.com>  Tue, 22 Jan 2014 17:27:25 +0400



libslave (1.0.5-1) UNRELEASED; urgency=low

  * Initial release. 

 -- alheio <zaharov.sy@gmail.com>  Tue, 10 Dec 2013 17:39:25 +0400


slave_log_event.h
/* Copyright 2011 ZAO "Begun".
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 3 of the License, or (at your option)
 * any later version.
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library.  If not, see <http://www.gnu.org/licenses/>.
*/


#ifndef __SLAVE_SLAVE_LOG_EVENT_H
#define __SLAVE_SLAVE_LOG_EVENT_H


#include "relayloginfo.h"


namespace slave {



/*
 *
 */

enum Log_event_type
{

  UNKNOWN_EVENT= 0,
  START_EVENT_V3= 1,
  QUERY_EVENT= 2,
  STOP_EVENT= 3,
  ROTATE_EVENT= 4,
  INTVAR_EVENT= 5,
  LOAD_EVENT= 6,
  SLAVE_EVENT= 7,
  CREATE_FILE_EVENT= 8,
  APPEND_BLOCK_EVENT= 9,
  EXEC_LOAD_EVENT= 10,
  DELETE_FILE_EVENT= 11,
  NEW_LOAD_EVENT= 12,
  RAND_EVENT= 13,
  USER_VAR_EVENT= 14,
  FORMAT_DESCRIPTION_EVENT= 15,
  XID_EVENT= 16,
  BEGIN_LOAD_QUERY_EVENT= 17,
  EXECUTE_LOAD_QUERY_EVENT= 18,

  TABLE_MAP_EVENT = 19,

  PRE_GA_WRITE_ROWS_EVENT = 20,
  PRE_GA_UPDATE_ROWS_EVENT = 21,
  PRE_GA_DELETE_ROWS_EVENT = 22,

  WRITE_ROWS_EVENT_V1 = 23,
  UPDATE_ROWS_EVENT_V1 = 24,
  DELETE_ROWS_EVENT_V1 = 25,

  INCIDENT_EVENT = 26,

  HEARTBEAT_LOG_EVENT= 27,

  // 5.6.x new events

  IGNORABLE_LOG_EVENT      = 28,
  ROWS_QUERY_LOG_EVENT     = 29,

  WRITE_ROWS_EVENT_V2      = 30,
  UPDATE_ROWS_EVENT_V2     = 31,
  DELETE_ROWS_EVENT_V2     = 32,

  GTID_LOG_EVENT           = 33,
  ANONYMOUS_GTID_LOG_EVENT = 34,

  PREVIOUS_GTIDS_LOG_EVENT = 35,

  ENUM_END_EVENT
};

#define LOG_EVENT_TYPES (ENUM_END_EVENT-1)


#define EVENT_TYPE_OFFSET        4
#define SERVER_ID_OFFSET         5
#define EVENT_LEN_OFFSET         9
#define LOG_POS_OFFSET           13

#define LOG_EVENT_HEADER_LEN     19

#define ROTATE_HEADER_LEN        8
#define R_POS_OFFSET             0

#define QUERY_HEADER_MINIMAL_LEN (4 + 4 + 1 + 2)
#define QUERY_HEADER_LEN         (QUERY_HEADER_MINIMAL_LEN + 2)
#define Q_STATUS_VARS_LEN_OFFSET 11
#define Q_DB_LEN_OFFSET          8

#define TABLE_MAP_HEADER_LEN     8
#define TM_MAPID_OFFSET          0

#define ROWS_HEADER_LEN_V1       8
#define RW_MAPID_OFFSET          0
#define ROWS_HEADER_LEN_V2       10

#define LOG_EVENT_MINIMAL_HEADER_LEN 19

#define ST_BINLOG_VER_LEN           2
#define ST_BINLOG_VER_OFFSET        0
#define ST_SERVER_VER_LEN 50
#define ST_SERVER_VER_OFFSET        (ST_BINLOG_VER_OFFSET + 2)
#define ST_CREATED_OFFSET     (ST_SERVER_VER_OFFSET + ST_SERVER_VER_LEN)
#define ST_COMMON_HEADER_LEN_OFFSET (ST_CREATED_OFFSET + 4)

#define START_V3_HEADER_LEN     (2 + ST_SERVER_VER_LEN + 4)


// binlog_checksum stuff (new in 5.6.x)

enum enum_binlog_checksum_alg {
    BINLOG_CHECKSUM_ALG_OFF      = 0,   // Events are without checksum though its generator
                                        // is checksum-capable New Master (NM).
    BINLOG_CHECKSUM_ALG_CRC32    = 1,   // CRC32 of zlib algorithm.
    BINLOG_CHECKSUM_ALG_ENUM_END,       // the cut line: valid arg range is [1, 0x7f].
    BINLOG_CHECKSUM_ALG_UNDEF    = 255  // special value to tag undetemined yet checksum
                                        // or events from checksum-unaware servers
};

#define CHECKSUM_CRC32_SIGNATURE_LEN    4
/**
    defined statically while there is just one alg implemented
*/
#define BINLOG_CHECKSUM_LEN             CHECKSUM_CRC32_SIGNATURE_LEN
#define BINLOG_CHECKSUM_ALG_DESC_LEN    1   /* 1 byte checksum alg descriptor */


//-----------------------------------------------------------------------------------------

struct Basic_event_info {

    slave::Log_event_type type;
    size_t log_pos;
    time_t when;
    unsigned int server_id;

    const char* buf;
    unsigned int event_len;

    Basic_event_info() : type(UNKNOWN_EVENT), log_pos(0), when(0), server_id(0), buf(NULL), event_len(0)
        {}

    void parse(const char* b, unsigned int el);
};

struct Rotate_event_info {

    unsigned int ident_len;
    std::string new_log_ident;
    size_t pos;

    Rotate_event_info(const char* buf, unsigned int event_len);
};

struct Query_event_info {

    std::string query;

    Query_event_info(const char* buf, unsigned int event_len);
};

struct Table_map_event_info {

    unsigned long m_table_id;
    std::string m_tblnam;
    std::string m_dbnam;

    Table_map_event_info(const char* buf, unsigned int event_len);
};

struct Row_event_info {

    unsigned long m_width;
    unsigned long m_table_id;

    std::vector<unsigned char> m_cols;
    std::vector<unsigned char> m_cols_ai;

    unsigned char* m_rows_buf;
    unsigned char* m_rows_end;

    bool has_after_image;

    Row_event_info(const char* buf, unsigned int event_len,
                   bool do_update, bool is_master_5_6_x);
};


bool read_log_event(const char* buf, unsigned int event_len, Basic_event_info& info,
                    EventStatIface* event_stat, bool is_master_5_6_x);

void apply_row_event(slave::RelayLogInfo& rli, const Basic_event_info& bei, const Row_event_info& roi, ExtStateIface &ext_state, EventStatIface* event_stat);


//------------------------------------------------------------------------------------------







} //namespace


#endif
